scheme/arm-cca/realm.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+
+package arm_cca
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/veraison/ccatoken"
+	"github.com/veraison/ear"
+	"github.com/veraison/services/handler"
+	"github.com/veraison/services/log"
+	"github.com/veraison/services/scheme/common/arm"
+	realm "github.com/veraison/services/scheme/common/cca/realm"
+)
+
+type CCARealmAttester struct {
+}
+
+func (CCARealmAttester) PerformAppraisal(
+	appraisal *ear.Appraisal,
+	claimsMap map[string]interface{},
+	endorsements []handler.Endorsement) error {
+
+	realmEndorsements := arm.FilterRefVal(endorsements, "realm.reference-value")
+	claims, err := realm.MapToRealmClaims(claimsMap)
+	if err != nil {
+		return fmt.Errorf("unable to get realm claims from Realm claims map: %w", err)
+	}
+
+	// If crypto verification (including chaining) completes correctly,
+	// we can safely assume the Realm instance to be trustworthy
+	appraisal.TrustVector.InstanceIdentity = ear.TrustworthyInstanceClaim
+	appraisal.TrustVector.Executables = ear.UnrecognizedRuntimeClaim
+
+	for _, endorsement := range realmEndorsements {
+		rimRpvMatch := matchRimRpv(claims, &endorsement)
+		if rimRpvMatch {
+			appraisal.TrustVector.Executables = ear.ApprovedBootClaim
+			// Match REM's
+			remMatch := matchREMs(claims, &endorsement)
+			if remMatch {
+				appraisal.TrustVector.Executables = ear.ApprovedRuntimeClaim
+			}
+			break
+		} else {
+			// TO DO “UNRECOGNIZED_BOOT" is missing
+			appraisal.TrustVector.Executables = ear.ContraindicatedRuntimeClaim
+		}
+	}
+	appraisal.UpdateStatusFromTrustVector()
+	appraisal.VeraisonAnnotatedEvidence = &claimsMap
+
+	return nil
+}
+
+func matchRimRpv(claims ccatoken.IClaims, endorsement *handler.Endorsement) bool {
+	// Compare RIM and PV (if provided)
+	rimClaim, err := claims.GetInitialMeasurement()
+	if err != nil {
+		log.Errorf("matchRimRpv failed: %w", handler.BadEvidence(err))
```

Comment:
These log messages do not seem to provide the information needed to correlate the error with the current appraisal.  Which makes them not very useful, apart from when debugging an isolated transaction.

Reply:
As discussed, we will raise a separate issue to track `end-to-end` appraisal sequence to distinguish concurrent appraisal procedures!

Reply:
See https://github.com/veraison/services/issues/236